--- gecko_driver.ori.js	2016-11-09 11:02:14.618459300 +0100
+++ gecko_driver.js	2016-11-09 11:02:23.733258100 +0100
@@ -31,7 +31,7 @@
         var version = this.version();
         // No 32-bit builds before 0.10.0
         if (semver.lte(version, '0.10.0')) {
-            if (arch === 'x64') {
+            if (arch !== 'x64') {
                 throw new Error('GeckoDriver doesn\'t support ${ostype} ${arch}!');
             }
         }
