--- a/preact-compat.js
+++ b/preact-compat.js
@@ -1,6 +1,6 @@
 (function (global, factory) {
-	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('prop-types'), require('preact')) :
-	typeof define === 'function' && define.amd ? define(['prop-types', 'preact'], factory) :
+	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('proptypes'), require('preact')) :
+	typeof define === 'function' && define.amd ? define(['proptypes', 'preact'], factory) :
 	(global.preactCompat = factory(global.PropTypes,global.preact));
 }(this, (function (PropTypes,preact) {
 
