// home/views/home/Step2View.js
import React from "%{ amd: 'react', brunch: '!React', common: 'react' }";
import ReactModelView from "umd-core/src/views/ReactModelView";
import ArrayComponent from "umd-core/src/components/ArrayComponent";

module.exports = class Step2View extends ReactModelView {
    get phones() {
        return [
            {
                name: "Nexus S",
                snippet: "Fast just got faster with Nexus S.",
                age: 1
            },
            {
                name: "Motorola XOOM™ with Wi-Fi",
                snippet: "The Next, Next Generation tablet.",
                age: 2
            },
            {
                name: "MOTOROLA XOOM™",
                snippet: "The Next, Next Generation tablet.",
                age: 3
            }
        ];
    }

    initialize() {
        // set order to 'age'
        this.inline.set("order", "age");
    }

    childNode(phone, index) {
        return (
            <li key={ index }>
                <span className="name">{ phone.name }</span>
                <p className="snippet">{ phone.snippet }</p>
            </li>
        );
    }

    render() {
        return (
            <div className="container-fluid">
                <div className="row">
                    <div className="col-md-4">
                        { /* component model properties are available with this.inline.get(name) */ }
                        Search: <input spModel="query" />

                        Sort by:
                        <select spModel="order">
                            <option value="name">Alphabetical</option>
                            <option value="age">Newest</option>
                        </select>
                    </div>
                    <div className="col-md-8">
                        <ArrayComponent
                            tagName="ul"
                            className="phones"
                            collection={ this.phones }
                            filter={ this.inline.get("query") }
                            order={ this.inline.get("order") }
                            childNode={ this.childNode }
                        />
                    </div>
                </div>
            </div>
        );
    }
};
