// home/views/ModalUserEditView.js
import React from "%{ amd: 'react', common: '!React' }";
import Dialog from "umd-core/src/components/Dialog";
import UserEditView from "./UserEditView";
import { i18n } from "umd-core/src/common";

const btnRaisedRipple = "mdl-button mdl-js-button mdl-button--raised mdl-js-ripple-effect";
const btnAccent = `${ btnRaisedRipple } mdl-button--accent`;

class ModalUserEditView extends UserEditView {
    constructor() {
        super(...arguments);
        this.onCancel = this.onCancel.bind(this);
        this.onDelete = this.onDelete.bind(this);
        this.onSave = this.onSave.bind(this);
        this.showModal = this.showModal.bind(this);
    }

    showModal(proxy) {
        const evt = proxy.nativeEvent;
        const which = evt.which;

        if (which !== 1 || evt.altKey || evt.ctrlKey || evt.shiftKey) {
            return;
        }

        proxy.preventDefault();
        this.refs.dialog.showModal();
    }

    onSave() {
        this.refs.dialog.close();
        ModalUserEditView.__super__.onSave.apply(this, arguments);
    }

    onDelete() {
        this.refs.dialog.close();
        ModalUserEditView.__super__.onDelete.apply(this, arguments);
    }

    onCancel() {
        this.refs.dialog.close();
    }

    render() {
        return (
            <span>
                {/*
                    React listen click event on the document.
                    However, the navigate handler on a[href] click is before the React handler.
                    For that reason, there is a class name to bypass navigate handler on a[href]
                */}
                <a href={ this.props.href } className={ `no-navigate ${ btnAccent }` } onClick={ this.showModal }>
                    { i18n.t(this.props.model.id ? "button.edit" : "label.user.create") }
                </a>
                <Dialog ref="dialog" className="modal">
                    <form onSubmit={ this.handleSubmit }>
                        <div className="mdl-dialog__title">{ this.getTitle() }</div>
                        <div className="mdl-dialog__content">{ this.getFieldSet() }</div>
                        <div className="mdl-dialog__actions">{ this.getButtons() }</div>
                    </form>
                </Dialog>
            </span>
        );
    }
}

ModalUserEditView.__super__ = Object.getPrototypeOf(ModalUserEditView.prototype);
module.exports = ModalUserEditView;
